sdk/java/dagger-java-annotation-processor/src/main/java/io/dagger/annotation/processor/DaggerModuleAnnotationProcessor.java

```diff
@@ -0,0 +1,454 @@
+                              .beginControlFlow(
+                                  "for ($T fnArg : fnArgs)", FunctionCallArgValue.class)
+                              .addStatement("inputArgs.put(fnArg.name(), fnArg.value())")
+                              .endControlFlow()
+                              .addCode("\n")
+                              .addStatement("$T result", JSON.class)
+                              .beginControlFlow("if (parentName.isEmpty())")
+                              .addStatement("$T modID = register()", ModuleID.class)
+                              .addStatement("result = $T.toJSON(modID)", JsonConverter.class)
+                              .nextControlFlow("else")
+                              .addStatement(
+                                  "result = invoke(parentJson, parentName, fnName, inputArgs)")
+                              .endControlFlow()
+                              .addStatement("fnCall.returnValue(result)")
+                              .nextControlFlow("catch ($T e)", Exception.class)
+                              .addStatement("fnCall.returnError(dag.error(e.getMessage()))")
+                              .addStatement("throw e")
+                              .endControlFlow()
+                              .build())
+                      .addMethod(rm.build())
+                      .addMethod(im.build())
+                      .build())
+              .addFileComment("This class has been generated by dagger-java-sdk. DO NOT EDIT.")
+              .indent("  ")
+              .build();
+
+      return f;
+    } catch (ClassNotFoundException e) {
+      throw new RuntimeException(e);
+    }
+  }
+
+  @Override
+  public boolean process(Set<? extends TypeElement> annotations, RoundEnvironment roundEnv) {
+    ModuleInfo moduleInfo = generateModuleInfo(annotations, roundEnv);
+
+    if (moduleInfo.objects().length == 0) {
+      return true;
+    }
+
+    try {
+      JavaFile f = generate(moduleInfo);
+
+      f.writeTo(processingEnv.getFiler());
+    } catch (IOException e) {
+      throw new RuntimeException(e);
+    }
+
+    return true;
+  }
+
+  static CodeBlock typeDef(String name) throws ClassNotFoundException {
+    if (name.equals("int")) {
+      return CodeBlock.of(
+          "dag.typeDef().withKind($T.$L)", TypeDefKind.class, TypeDefKind.INTEGER_KIND.name());
+    } else if (name.equals("boolean")) {
+      return CodeBlock.of(
+          "dag.typeDef().withKind($T.$L)", TypeDefKind.class, TypeDefKind.BOOLEAN_KIND.name());
+    } else if (name.startsWith("java.util.List<")) {
+      name = name.substring("java.util.List<".length(), name.length() - 1);
+      return CodeBlock.of("dag.typeDef().withListOf($L)", typeDef(name).toString());
+    } else if (name.endsWith("[]")) {
+      name = name.substring(0, name.length() - 2);
+      return CodeBlock.of("dag.typeDef().withListOf($L)", typeDef(name).toString());
+    }
+
+    var clazz = Class.forName(name);
```

Comment:
This statement may throw a `ClassNotFoundException` when having a function returning itself or another module object since these objects have not yet been compiled, and then the classes cannot be loaded into the JVM.

For instance:
```
@Object
class Foo() extends Base {
  private Container ctr;
 
  public Foo() {
    this.ctr = dag.container();
  }
 
  public Foo WithAlpine() {
    this.ctr = this.ctr.from("alpine:latest")
    return this;
}
```

Additional type metadata (for instance if the type is an enum or not) should be parsed from the AST provided by the annotation parser rather than introspection

Reply:
Hum, true.
I'll improve that.

Regarding enums, for now I haven't found a way to check if it's an enum from the annotation parser 🤔 Maybe I'm missing something.

But at least I can handle the class not found correctly, especially as this code is only used for enums.

Reply:
Unfortunately the same issue exists also in the `classForName(TypeInfo ti)` method

Reply:
I just improved that part too.
And I expanded the test case to cover to reference the module object itself, and more regarding primitive types and lists.